docs/dev/relay_add_pass.rst

```diff
@@ -0,0 +1,125 @@
+
+.. code:: c
+
+    void VisitExpr_(const ConstantNode* n) final {
+      memo_[GetRef<Constant>(n)] = true;
+    }
+
+    void VisitExpr_(const TupleNode* n) final {
+      bool result = true;
+      for (const auto& field : n->fields) {
+        if (!Check(field)) {
+          result = false;
+          break;
+        }
+      }
+      memo_[GetRef<Tuple>(n)] = result;
+    }
+
+Note that we don't need to define a ``VisitExpr_`` function for every type of
+node, because ``ExprVisitor`` defines a default implementation for every
+node. This default implementation simply visits all of its fields that are
+nodes. So we can default to returning false and only define ``VisitExpr_``
+methods for the node types that could possibly change the result to true
+(i.e., the ones listed in our constness condition earlier).
+
+Writing an Expression Mutator
+-----------------------------
+
+The second component of the constant folding pass is the ``ConstantFolder``,
+which extends ``ExprMutator`` and internally uses ``ConstantChecker``. The
+``ConstantFolder`` class is used to implement the ``FoldConstant`` function,
+so it can be registered to become an API endpoint.
+
+In the ``ConstantChecker``, we defined our own ``memo_`` field, but we don't
+need to do the same in ``ConstantFolder``. ``ExprMutator`` already provides a
+memoizer, because we know its ``VisitExpr_`` functions will always take
+expressions and return expressions.
+
+Similarly to ``ExprVisitor``, ``ExprMutator`` also defines default
+implementations for all node types. Each ``VisitExpr_`` method visits all of
+its node fields and creates a new node if any of the resulting fields have
+changed. Otherwise, the original node is returned.
+
+In Relay, there are three types of nodes that we want to perform constant
+folding on:
+
+- ``LetNode``s
+- ``TupleItemGetNode``s
+- ``CallNode``s
+
+For ``LetNode``s, we first attempt to const-fold the value being bound in the
+expression. If we can, then we populate ``memo_`` and return the result of
+visiting the body---essentially, propagating the bound value to its use sites
+in the body. If we can't const-fold the bound value, we mimic the default
+implementation.
+
+For ``TupleItemGetNode``s, we check if the index expression can be
+const-folded. If so, we replace the tuple get with the field the evaluated
+constant points to.
+
+For ``CallNode``s, we first use the ``VisitExpr_`` of ``ExprMutator`` to
+visit the call, which const-folds all of the call parameters. Then we
+evaluate the call only if all of the arguments are constant (using the
+``ConstantChecker``). Evaluating the call produces a value, so we use a
+helper method ``ValueToExpr`` to allow us to place the evaluated expression
+back into the AST.
+
+Wrapping the API in Python
+-----------------------------
+
+TODO: The content of this section should be almost identical to the
+corresponding section in ``docs/dev/relay_add_op.rst``. Can we just link to
+there?
+
+Summary
```

Comment:
This summary seems unnecessary since it does not recap the points from the different sections (nor do I believe that should be at the end; a high-level summary would be best for the beginning).